src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs

```diff
@@ -814,6 +821,38 @@ private void CalculateExcludedLibraries()
             {
                 _dependencyLibraries[packageToExcludeFromRuntime].ExcludeFromRuntime = true;
             }
+
+            //  Include transitive dependencies of all top-level dependencies
+            Dictionary<string, DependencyLibrary> includedDependencies = new(StringComparer.OrdinalIgnoreCase);
+            Stack<string> dependencyListToWalk = new(_mainProjectDependencies);
+
+            while (dependencyListToWalk.Count != 0)
+            {
+                var dependencyName = dependencyListToWalk.Pop();
+                if (!includedDependencies.ContainsKey(dependencyName) && _excludeFromPublishPackageIds?.Contains(dependencyName) != true)
+                {
+                    //  There may not be a library in the assets file if a referenced project has
+                    //  PrivateAssets="all" for a package reference, and there is a package in the graph
+                    //  that depends on the same package.
+                    if (_dependencyLibraries.TryGetValue(dependencyName, out var dependencyLibrary))
+                    {
+                        includedDependencies.Add(dependencyName, dependencyLibrary);
+                        foreach (var newDependency in _libraryDependencies[dependencyName])
+                        {
+                            dependencyListToWalk.Push(newDependency.Name);
+                        }
+                    }
+                }
```

Comment:
nit: Nested boolean conditions

```suggestion
                //  There may not be a library in the assets file if a referenced project has
                //  PrivateAssets="all" for a package reference, and there is a package in the graph
                //  that depends on the same package.
                if (!includedDependencies.ContainsKey(dependencyName) &&
                    _excludeFromPublishPackageIds?.Contains(dependencyName) != true &&
                    _dependencyLibraries.TryGetValue(dependencyName, out var dependencyLibrary))
                {
                    includedDependencies.Add(dependencyName, dependencyLibrary);
                    foreach (var newDependency in _libraryDependencies[dependencyName])
                    {
                        dependencyListToWalk.Push(newDependency.Name);
                    }
                }
```